posthog/hogql/grammar/HogQLParser.g4

```diff
@@ -0,0 +1,228 @@
+    from=fromClause?
+    arrayJoinClause?
+    windowClause?
+    prewhereClause?
+    where=whereClause?
+    groupByClause? (WITH (CUBE | ROLLUP))? (WITH TOTALS)?
+    havingClause?
+    orderByClause?
+    limitByClause?
+    limitClause?
+    settingsClause?
+    ;
+
+withClause: WITH columnExprList;
+topClause: TOP DECIMAL_LITERAL (WITH TIES)?;
+fromClause: FROM joinExpr;
+arrayJoinClause: (LEFT | INNER)? ARRAY JOIN columnExprList;
+windowClause: WINDOW identifier AS LPAREN windowExpr RPAREN;
+prewhereClause: PREWHERE columnExpr;
+whereClause: WHERE columnExpr;
+groupByClause: GROUP BY ((CUBE | ROLLUP) LPAREN columnExprList RPAREN | columnExprList);
+havingClause: HAVING columnExpr;
+orderByClause: ORDER BY orderExprList;
+projectionOrderByClause: ORDER BY columnExprList;
+limitByClause: LIMIT limitExpr BY columnExprList;
+limitClause: LIMIT limitExpr (WITH TIES)?;
+settingsClause: SETTINGS settingExprList;
+
+joinExpr
+    : joinExpr (GLOBAL | LOCAL)? joinOp? JOIN joinExpr joinConstraintClause  # JoinExprOp
+    | joinExpr joinOpCross joinExpr                                          # JoinExprCrossOp
+    | tableExpr FINAL? sampleClause?                                         # JoinExprTable
+    | LPAREN joinExpr RPAREN                                                 # JoinExprParens
+    ;
+joinOp
+    : ((ALL | ANY | ASOF)? INNER | INNER (ALL | ANY | ASOF)? | (ALL | ANY | ASOF))  # JoinOpInner
+    | ( (SEMI | ALL | ANTI | ANY | ASOF)? (LEFT | RIGHT) OUTER?
+      | (LEFT | RIGHT) OUTER? (SEMI | ALL | ANTI | ANY | ASOF)?
+      )                                                                             # JoinOpLeftRight
+    | ((ALL | ANY)? FULL OUTER? | FULL OUTER? (ALL | ANY)?)                         # JoinOpFull
+    ;
+joinOpCross
+    : (GLOBAL|LOCAL)? CROSS JOIN
+    | COMMA
+    ;
+joinConstraintClause
+    : ON columnExprList
+    | USING LPAREN columnExprList RPAREN
+    | USING columnExprList
+    ;
+
+sampleClause: SAMPLE ratioExpr (OFFSET ratioExpr)?;
+limitExpr: columnExpr ((COMMA | OFFSET) columnExpr)?;
+orderExprList: orderExpr (COMMA orderExpr)*;
+orderExpr: columnExpr (ASCENDING | DESCENDING | DESC)? (NULLS (FIRST | LAST))? (COLLATE STRING_LITERAL)?;
+ratioExpr: numberLiteral (SLASH numberLiteral)?;
+settingExprList: settingExpr (COMMA settingExpr)*;
+settingExpr: identifier EQ_SINGLE literal;
+
+windowExpr: winPartitionByClause? winOrderByClause? winFrameClause?;
+winPartitionByClause: PARTITION BY columnExprList;
+winOrderByClause: ORDER BY orderExprList;
+winFrameClause: (ROWS | RANGE) winFrameExtend;
+winFrameExtend
+    : winFrameBound                             # frameStart
+    | BETWEEN winFrameBound AND winFrameBound   # frameBetween
+    ;
+winFrameBound: (CURRENT ROW | UNBOUNDED PRECEDING | UNBOUNDED FOLLOWING | numberLiteral PRECEDING | numberLiteral FOLLOWING);
+//rangeClause: RANGE LPAREN (MIN identifier MAX identifier | MAX identifier MIN identifier) RPAREN;
+
+
+
+// Columns
```

Comment:
Most column syntax from here is already supported. The exceptions are arrays and lambdas.